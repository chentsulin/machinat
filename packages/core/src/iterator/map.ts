import type { MachinatNode, MachinatRenderable } from '../types';
import type { NodeMapper } from './types';
import traverse from './traverse';

type MapTraverseContext<Mapped, Payload> = {
  mappedArray: Array<Mapped>;
  mapper: NodeMapper<Mapped, Payload>;
  payload: Payload;
};

/** @internal */
const mapCallback = <Mapped, Payload>(
  child: MachinatRenderable,
  path: string,
  context: MapTraverseContext<Mapped, Payload>
) => {
  const { mapper, payload, mappedArray } = context;
  const mapped = mapper(child, path, payload);
  mappedArray.push(mapped);
};

const map = <Mapped, Payload>(
  children: MachinatNode,
  mapper: NodeMapper<Mapped, Payload>,
  prefix: string,
  payload: Payload
): Mapped[] | null | undefined => {
  if (children === undefined || children === null) {
    return children;
  }

  const context: MapTraverseContext<Mapped, Payload> = {
    mapper,
    payload,
    mappedArray: [],
  };

  traverse(children, prefix, context, mapCallback);
  return context.mappedArray;
};

export default map;
