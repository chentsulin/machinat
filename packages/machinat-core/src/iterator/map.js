// @flow
import type { MachinatNode } from '../types';
import type { NodeMapper } from './types';
import traverse from './traverse';

type MapTraverseContext<Mapped, Payload> = {
  mappedArray: Array<Mapped>,
  mapper: NodeMapper<Mapped, Payload>,
  payload: Payload,
};

const mapCallback = <Mapped, Payload>(
  child,
  path,
  context: MapTraverseContext<Mapped, Payload>
) => {
  const { mapper, payload, mappedArray } = context;
  const mapped = mapper(child, path, payload);
  mappedArray.push(mapped);
};

const map = <Mapped, Payload>(
  children: MachinatNode,
  mapper: NodeMapper<Mapped, Payload>,
  prefix: string,
  payload: any
): ?Array<Mapped> => {
  if (children === undefined || children === null) {
    return children;
  }

  const context: MapTraverseContext<Mapped, Payload> = {
    mapper,
    payload,
    mappedArray: [],
  };

  traverse(children, prefix, context, mapCallback);
  return context.mappedArray;
};

export default map;
