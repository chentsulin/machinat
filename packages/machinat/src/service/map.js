// @flow
import type { Interfaceable } from './types';

type ProvisionBranches<T> = Map<
  Interfaceable,
  {| default: null | T, branches: null | Map<string, T> |}
>;

export default class ProvisionMap<T> {
  _mapping: ProvisionBranches<T>;

  constructor(mapping?: ProvisionBranches<T>) {
    this._mapping = mapping || new Map();
  }

  getDefault(target: Interfaceable): null | T {
    const mapped = this._mapping.get(target);
    return mapped ? mapped.default : null;
  }

  getPlatform(target: Interfaceable, platform: string): null | T {
    const mapped = this._mapping.get(target);
    return mapped && mapped.branches
      ? mapped.branches.get(platform) || null
      : null;
  }

  setDefault(target: Interfaceable, value: T): boolean {
    const mapped = this._mapping.get(target);
    if (mapped) {
      const isUpdating = !!mapped.default;
      mapped.default = value;
      return isUpdating;
    }

    this._mapping.set(target, { default: value, branches: null });
    return false;
  }

  setPlatform(target: Interfaceable, platform: string, value: T): boolean {
    const mapped = this._mapping.get(target);
    if (!mapped) {
      const branches = new Map();
      branches.set(platform, value);

      this._mapping.set(target, { default: null, branches });
      return false;
    }

    if (!mapped.branches) {
      mapped.branches = new Map().set(platform, value);
      return false;
    }

    const { branches } = mapped;
    const isUpdating = branches.has(platform);
    branches.set(platform, value);

    return isUpdating;
  }

  merge(concatee: ProvisionMap<T>): ProvisionMap<T> {
    for (const [target, provided] of concatee._mapping) {
      if (provided.default) {
        this.setDefault(target, provided.default);
      }

      if (provided.branches) {
        for (const [platform, value] of provided.branches) {
          this.setPlatform(target, platform, value);
        }
      }
    }

    return this;
  }

  /* :: @@iterator(): Generator<[Interfaceable, void | string, any], void, void> {return ({}: any)} */
  *[Symbol.iterator](): Generator<
    [Interfaceable, void | string, any],
    void,
    void
  > {
    for (const [target, provided] of this._mapping) {
      if (provided.default) {
        yield [target, undefined, provided.default];
      }

      if (provided.branches) {
        for (const [platform, value] of provided.branches) {
          yield [target, platform, value];
        }
      }
    }
  }
}
