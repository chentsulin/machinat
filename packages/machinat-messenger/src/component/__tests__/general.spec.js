import Machinat from '@machinat/core';
import map from '@machinat/core/iterator/map';
import generalDelegate from '../general';

const render = async node => {
  const renderings = map(
    node,
    (element, path) =>
      typeof element === 'string'
        ? { type: 'text', value: element, node: element, path }
        : typeof element.type === 'string'
        ? generalDelegate(element, render, path)
        : { value: { non: 'text' }, node: element, path },
    '$'
  );
  return renderings ? [].concat(...(await Promise.all(renderings))) : null;
};

test('shallow textual elements match snapshot', async () => {
  await expect(
    render(
      <>
        <text>abc</text>
        <b>important</b>
        <i>italic</i>
        <del>nooooo</del>
        <br />
        <code>foo.bar()</code>
        <pre>foo.bar('hello world')</pre>
      </>
    )
  ).resolves.toMatchSnapshot();
});

test('nested textual elements match snapshot', async () => {
  const segments = await render(
    <text>
      123{' '}
      <code>
        Hello, <b>Luke Skywalker!</b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        I'm your <del>FATHER</del> <code>droid</code>.
      </i>
      <br />
      <br />
      <pre>May the force be with you!</pre> abc
    </text>
  );
  expect(segments).toMatchSnapshot();
});

test('with break placed in children', async () => {
  const children = (
    <>
      foo
      <br />
      bar
    </>
  );

  const segments = await render([
    <text>{children}</text>,
    <b>{children}</b>,
    <i>{children}</i>,
    <del>{children}</del>,
    <code>{children}</code>,
    <pre>{children}</pre>,
  ]);
  expect(segments).toMatchSnapshot();
});

test('should throw if non string value rendered', async () => {
  const NonText = () => {};
  const children = (
    <>
      foo
      <NonText />
      bar
    </>
  );

  for (const node of [
    <b>{children}</b>,
    <i>{children}</i>,
    <del>{children}</del>,
    <text>{children}</text>,
    <code>{children}</code>,
    <pre>{children}</pre>,
  ]) {
    // eslint-disable-next-line no-await-in-loop
    await expect(render(node)).rejects.toThrow(
      '<NonText /> at $::1 is not valid textual content'
    );
  }
});

test('media elements match snapshot', async () => {
  await expect(
    render([
      <img src="http://avatar.my.bot" />,
      <video src="http://vid.my.bot" />,
      <audio src="http://sound.my.bot" />,
      <file src="http://profile.my.bot" />,
    ])
  ).resolves.toMatchSnapshot();
});

test('return null if content is empty', async () => {
  for (const element of [<b />, <i />, <del />, <text />, <code />, <pre />]) {
    // eslint-disable-next-line no-await-in-loop
    await expect(render(element)).resolves.toEqual([null]);
  }
});
