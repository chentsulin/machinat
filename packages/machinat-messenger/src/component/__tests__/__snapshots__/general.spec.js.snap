// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media elements match snapshot 1`] = `
Array [
  Object {
    "node": <img
      src="http://avatar.my.bot"
    />,
    "path": "$:0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://avatar.my.bot",
          },
          "type": "image",
        },
      },
    },
  },
  Object {
    "node": <video
      src="http://vid.my.bot"
    />,
    "path": "$:1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://vid.my.bot",
          },
          "type": "video",
        },
      },
    },
  },
  Object {
    "node": <audio
      src="http://sound.my.bot"
    />,
    "path": "$:2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://sound.my.bot",
          },
          "type": "audio",
        },
      },
    },
  },
  Object {
    "node": <file
      src="http://profile.my.bot"
    />,
    "path": "$:3",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "url": "http://profile.my.bot",
          },
          "type": "file",
        },
      },
    },
  },
]
`;

exports[`nested textual elements match snapshot 1`] = `
Array [
  Object {
    "node": <text>
      123
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        I'm your 
        <del>
          FATHER
        </del>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       abc
    </text>,
    "path": "$",
    "type": "text",
    "value": "123 \`Hello, *Luke Skywalker!*\`",
  },
  Object {
    "node": <br />,
    "path": "$:3",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <text>
      123
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        I'm your 
        <del>
          FATHER
        </del>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       abc
    </text>,
    "path": "$",
    "type": "text",
    "value": "You know what?",
  },
  Object {
    "node": <br />,
    "path": "$:5",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <text>
      123
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        I'm your 
        <del>
          FATHER
        </del>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       abc
    </text>,
    "path": "$",
    "type": "text",
    "value": "_I'm your ~FATHER~ \`droid\`._",
  },
  Object {
    "node": <br />,
    "path": "$:7",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <br />,
    "path": "$:8",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <text>
      123
       
      <code>
        Hello, 
        <b>
          Luke Skywalker!
        </b>
      </code>
      <br />
      You know what?
      <br />
      <i>
        I'm your 
        <del>
          FATHER
        </del>
         
        <code>
          droid
        </code>
        .
      </i>
      <br />
      <br />
      <pre>
        May the force be with you!
      </pre>
       abc
    </text>,
    "path": "$",
    "type": "text",
    "value": "\`\`\`
May the force be with you!
\`\`\` abc",
  },
]
`;

exports[`shallow textual elements match snapshot 1`] = `
Array [
  Object {
    "node": <text>
      abc
    </text>,
    "path": "$::0",
    "type": "text",
    "value": "abc",
  },
  Object {
    "node": <b>
      important
    </b>,
    "path": "$::1",
    "type": "text",
    "value": "*important*",
  },
  Object {
    "node": <i>
      italic
    </i>,
    "path": "$::2",
    "type": "text",
    "value": "_italic_",
  },
  Object {
    "node": <del>
      nooooo
    </del>,
    "path": "$::3",
    "type": "text",
    "value": "~nooooo~",
  },
  Object {
    "node": <br />,
    "path": "$::4",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <code>
      foo.bar()
    </code>,
    "path": "$::5",
    "type": "text",
    "value": "\`foo.bar()\`",
  },
  Object {
    "node": <pre>
      foo.bar('hello world')
    </pre>,
    "path": "$::6",
    "type": "text",
    "value": "\`\`\`
foo.bar('hello world')
\`\`\`",
  },
]
`;

exports[`with break placed in children 1`] = `
Array [
  Object {
    "node": <text>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </text>,
    "path": "$:0",
    "type": "text",
    "value": "foo",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <text>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </text>,
    "path": "$:0",
    "type": "text",
    "value": "bar",
  },
  Object {
    "node": <b>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </b>,
    "path": "$:1",
    "type": "text",
    "value": "*foo*",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <b>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </b>,
    "path": "$:1",
    "type": "text",
    "value": "*bar*",
  },
  Object {
    "node": <i>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </i>,
    "path": "$:2",
    "type": "text",
    "value": "_foo_",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <i>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </i>,
    "path": "$:2",
    "type": "text",
    "value": "_bar_",
  },
  Object {
    "node": <del>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </del>,
    "path": "$:3",
    "type": "text",
    "value": "~foo~",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <del>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </del>,
    "path": "$:3",
    "type": "text",
    "value": "~bar~",
  },
  Object {
    "node": <code>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </code>,
    "path": "$:4",
    "type": "text",
    "value": "\`foo\`",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <code>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </code>,
    "path": "$:4",
    "type": "text",
    "value": "\`bar\`",
  },
  Object {
    "node": <pre>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </pre>,
    "path": "$:5",
    "type": "text",
    "value": "\`\`\`
foo
\`\`\`",
  },
  Object {
    "node": <br />,
    "path": "$::1",
    "type": "break",
    "value": undefined,
  },
  Object {
    "node": <pre>
      <Machinat.Fragment>
        foo
        <br />
        bar
      </Machinat.Fragment>
    </pre>,
    "path": "$:5",
    "type": "text",
    "value": "\`\`\`
bar
\`\`\`",
  },
]
`;
